py/sentry_streams/runner.py

```diff
@@ -1,9 +1,54 @@
 import os
 import sys
-from typing import Any, cast
+from typing import Any, MutableMapping, cast
 
 from pyflink.datastream import StreamExecutionEnvironment
-from sentry_streams.sinks import Pipeline, WithInput
+from sentry_streams.adapters.stream_adapter import StreamAdapter
+from sentry_streams.flink.flink_adapter import FlinkAdapter
+from sentry_streams.pipeline import (
+    Pipeline,
+    RuntimeTranslator,
+    WithInput,
+)
+
+
+# Traverses over edges in a PipelineGraph, building the
+# stream incrementally by applying steps and transformations
+# It currently has the structure to deal with, but has no
+# real support for, fan-in and fan-out streams
+def iterate_edges(
+    step_streams: MutableMapping[str, Any], p_graph: Pipeline, translator: RuntimeTranslator
+) -> Any:
+
+    output_stream = None
+    while step_streams:
+        for input_name in list(step_streams):
+            output_steps = p_graph.outgoing_edges[input_name]
+            input_stream = step_streams.pop(input_name)
+
+            if not output_steps:
+                continue
+
+            # check if the inputs are fanning out
+            if len(output_steps) > 1:
+                pass
+
+            else:
+                output_step_name = output_steps.pop()
+
+                # check if the inputs are fanning in
+                if len(p_graph.incoming_edges[output_step_name]) > 1:
+                    pass
+
+                # 1:1 between input and output stream
+                else:
+                    next_step: WithInput = cast(WithInput, p_graph.steps[output_step_name])
+                    print(f"Apply step: {next_step.name}")
+                    next_step_stream = translator.translate_step(next_step, input_stream)
+                    step_streams[next_step.name] = next_step_stream
+                    output_stream = next_step_stream
+
+    return output_stream
```

Comment:
Why do you need to return this ?

Reply:
There's kind of a small reason for returning it, I wanted to assert types of the input and output stream in test_pipeline. In order to do that and in order to be able to import `iterate_edges` I would need to return the output stream. I agree that it has no use outside of tests. 

I guess I can bake the assertion into the function `iterate_edges` itself, but it also seemed like a strange option considering the fact that the runner is then basically the test.

Reply:
fixed